File: Callback.swift
//
//  Callback.swift
//  RCTAgora
//
//  Created by LXH on 2020/4/9.
//  Copyright © 2020 Syan. All rights reserved.
//

import AgoraRtcKit
import Foundation

@objc
protocol Callback: class {
    func success(_ data: Any?)

    func failure(_ code: String, _ message: String)
}

extension Callback {
    func code(_ code: Int32?, _ runnable: ((Int32?) -> Any?)? = nil) {
        if code == nil || code! < 0 {
            let newCode = abs(Int(code ?? Int32(AgoraErrorCode.notInitialized.rawValue)))
            failure(String(newCode), AgoraRtcEngineKit.getErrorDescription(newCode) ?? "")
            return
        }

        let res = runnable?(code)
        success(res)
    }

    func resolve<T>(_ source: T?, _ runnable: (T) -> Any?) {
        guard let source = source else {
            let code = AgoraErrorCode.notInitialized.rawValue
            failure(String(code), AgoraRtcEngineKit.getErrorDescription(code) ?? "")
            return
        }

        let res = runnable(source)
        success(res)
    }
}

File: Callback.kt
package io.agora.rtc.base

import io.agora.rtc.Constants
import io.agora.rtc.RtcEngine
import kotlin.math.abs

abstract class Callback<T> {
    fun code(code: Int?) {
        val newCode = code ?: Constants.ERR_NOT_INITIALIZED
        if (newCode == 0) {
            success(null)
        } else if (newCode < 0) {
            failure(newCode.toString(), RtcEngine.getErrorDescription(abs(newCode)))
        }
    }

    abstract fun success(data: T?)

    abstract fun failure(code: String, message: String)
}
